Source/NSString.m

```diff
@@ -6248,6 +6256,164 @@ - (BOOL) containsString: (NSString *)string
+      while (YES)
+        {
+          // Since all characters are in a composed character sequence, enclosingRange == substringRange
+          NSRange enclosingRange = [self rangeOfComposedCharacterSequenceAtIndex: currentLocation];
+          CALL_BLOCK(block, 
+            substringNotRequired ? nil : [self substringWithRange: enclosingRange],
+            enclosingRange,
+            enclosingRange,
+            &stop);
+          if(stop) break;
+          currentLocation = enclosingRange.location + enclosingRange.length;
+        }
+    } 
+  else if (substringType == NSStringEnumerationByWords || substringType == NSStringEnumerationBySentences)
+    {
+      #if GS_USE_ICU
+      // These macros may be useful elsewhere.
+      #define GS_U_HANDLE_ERROR(errorCode, description) do { \
+        if (U_FAILURE(errorCode)) { \
+          NSWarnMLog(@"Error " description ": %s", u_errorName(errorCode)); \
+          return; \
+        } else if (errorCode < U_ZERO_ERROR) { \
+          NSWarnMLog(@"Warning " description ": %s", u_errorName(errorCode)); \
+        } \
+        errorCode = U_ZERO_ERROR; \
+      } while (NO)
+
+      BOOL byWords = substringType == NSStringEnumerationByWords;
+      NSUInteger length = range.length;
+      UChar characters[length];
+      [self getCharacters: characters range: range];
+      UErrorCode errorCode = U_ZERO_ERROR;
+      const char* locale = localized
+        ? [[[[NSLocale currentLocale] 
+                      localeIdentifier] 
+                      // @ss=standard will use lists of common abbreviations, such as Mr., Mrs., etc.
+                      stringByAppendingString: @"@ss=standard"] 
+                      UTF8String]
+        : "en_US_POSIX";
+      UBreakIterator* breakIterator = ubrk_open(byWords ? UBRK_WORD : UBRK_SENTENCE, // type
+                                                locale, // locale
+                                                characters, // text
+                                                length, // textLength
+                                                &errorCode);
+      GS_U_HANDLE_ERROR(errorCode, @"opening ICU break iterator");
+      ubrk_first(breakIterator);
+      while (YES)
+        {
+          // Make sure it's a valid substring.
+          BOOL isValidSubstring = YES;
+          
+          if (byWords) 
+            {
+              int32_t ruleStatus = ubrk_getRuleStatus(breakIterator);
+              // From ICU User Guide:
+              //   A status value UBRK_WORD_NONE indicates that the boundary does
+              //   not start a word or number.
+              // However, valid words seem to be UBRK_WORD_NONE, and invalid words
+              // seem to be UBRK_WORD_NONE_LIMIT.
+              isValidSubstring = ruleStatus != UBRK_WORD_NONE_LIMIT;
+              NSLog(@"Status for position %d (%d): %d", (int)currentLocation, (int)ubrk_current(breakIterator), (int) ruleStatus);
```

Comment:
Are you planing to leave these log statements in? In that case you should turn them into debug logs.